### ISSUE TITLE
RecursionError with recursive pydantic model and a default value

### ISSUE THREAD

**Issue by apirogov** [HypothesisWorks/hypothesis#3525]
Minimal working example:

```python
from pydantic import BaseModel
from hypothesis import given, strategies as st

class Dummy1(BaseModel):
    x: List[Dummy1] = []

Dummy1.update_forward_refs()


@given(st.from_type(Dummy1))
def test_modelgen(obj):
    print(obj.json(indent=2))
```

results in:

```
[...]
----
target = just([])

    def forced_value(target):
        try:
            return getattr(target, force_key)
        except AttributeError:
>           return getattr(target, cache_key)
E           RecursionError: maximum recursion depth exceeded while calling a Python object

/.../py3.8/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py:105: RecursionError
```

but without the default value (empty list) it works fine.

Python 3.8
pydantic 1.10.2
pytest 7.0.1
hypothesis 6.58

**Comment by Zac-HD** [comment]
A very similar issue was fixed in https://github.com/HypothesisWorks/hypothesis/pull/3656/, released as [version 6.75.9](https://hypothesis.readthedocs.io/en/latest/changes.html#v6-75-9), so I suspect that this is now a problem in Pydantic's Hypothesis plugin.

**Comment by jobh** [comment]
> A very similar issue was fixed in #3656, released as [version 6.75.9]

Actually it seems more related to #3671. When wrapping the recursive type in a list, the required recursion to validate the strategy is roughly doubled, and the default argument roughly doubles it again (due to the `one_of(just(defaultval, ...)`).

Effectively, this construction requires a recursionlimit of about 1100 to succeed on CPy. We can easily bump it up to that, but where's the limit....?

**Comment by jobh** [comment]
I spoke too quickly; there's a real issue here, I know what it is and will throw up a PR soon.
